mlflow/store/artifact/utils/models.py

```diff
@@ -63,6 +64,23 @@ def _parse_model_uri(uri):
     if parsed.scheme != "models":
         raise MlflowException(_improper_model_uri_msg(uri))
     path = parsed.path
+    # NB: If the model name contains a hash (#) in the name (historically allowed), reconstruct
+    # the path by adding the fragment identifier back into the parsed path.
+    # Multiple fragments will be included in the parsed fragment element.
+    fragment = f"#{parsed.fragment}" if parsed.fragment else ""
+    # If a query key (?) is included in the path, reconstitute the historically permitted
+    # character in a model name by adding the query identifier back into the parsed path.
+    # Multiple query identifiers (question marks) are included in the query; only the first
+    # query key in the path will be marked as the start of a query
+    query = f"?{parsed.query}" if parsed.query else ""
+    fragment_idx = uri.index(parsed.fragment) if parsed.fragment else len(uri)
+    query_idx = uri.index(parsed.query) if parsed.query else len(uri)
```

Comment:
What happens if the fragment content also appears in the path? Similarly, what happens if the query content also appears in the path?

I think doing `uri.index(query)` with query as defined on line 75 should work, but we should add test coverage for these cases.